File: src/execgo_playground/benchmarks/metrics.py
from __future__ import annotations

from typing import Any

from ..models import RunMetrics, StandardPlan, StageExecutionResult, VerifierResult


def compute_metrics(
    *,
    plan: StandardPlan,
    stage_results: list[StageExecutionResult],
    verdict: VerifierResult,
    wall_time_ms: int,
    artifact_hash_match: bool,
    recovery_expected: bool,
) -> RunMetrics:
    all_tasks = [task for stage in stage_results for task in stage.tasks]
    runtime_failures = sum(1 for task in all_tasks if task.status == "failed")
    retries = 0
    timeouts = 0
    for task in all_tasks:
        runtime = task.raw.get("runtime") or {}
        retries += max(0, int(runtime.get("attempt", 1)) - 1)
        error = runtime.get("error") or {}
        if error.get("code") == "timeout":
            timeouts += 1

    accepted = sum(int(stage.submission.get("accepted", 0)) for stage in stage_results)
    expected_accepts = sum(len(stage.task_ids) for stage in stage_results) or 1
    invalid_actions = len(plan.normalization_warnings)
    if any(stage.error for stage in stage_results):
        invalid_actions += 1

    return RunMetrics(
        plan_validity=1.0,
        submit_accept_rate=accepted / expected_accepts,
        scenario_success=1.0 if verdict.passed else 0.0,
        recovery_success=1.0 if (verdict.passed and recovery_expected) or not recovery_expected else 0.0,
        wall_time_ms=wall_time_ms,
        stage_count=len(plan.stages),
        task_count=len(all_tasks),
        retry_count=retries,
        timeout_count=timeouts,
        runtime_failure_count=runtime_failures,
        invalid_action_count=invalid_actions,
        determinism_drift=0.0,
        artifact_hash_match=1.0 if artifact_hash_match else 0.0,
    )

File: src/execgo_playground/benchmarks/metrics.ts
import type { RunMetrics, StageExecutionResult, StandardPlan, TaskSnapshot, VerifierResult } from "../models.js";

function runtimePayload(task: TaskSnapshot): Record<string, unknown> {
  const rawRuntime = task.raw.runtime;
  return rawRuntime && typeof rawRuntime === "object" && !Array.isArray(rawRuntime) ? (rawRuntime as Record<string, unknown>) : {};
}

export function computeMetrics(input: {
  plan: StandardPlan;
  stageResults: StageExecutionResult[];
  verdict: VerifierResult;
  wallTimeMs: number;
  artifactHashMatch: boolean;
  recoveryExpected: boolean;
}): RunMetrics {
  const allTasks = input.stageResults.flatMap((stage) => stage.tasks);
  const runtimeFailures = allTasks.filter((task) => task.status === "failed").length;
  let retries = 0;
  let timeouts = 0;
  for (const task of allTasks) {
    const runtime = runtimePayload(task);
    retries += Math.max(0, Number(runtime.attempt ?? 1) - 1);
    const error = runtime.error && typeof runtime.error === "object" ? (runtime.error as Record<string, unknown>) : {};
    if (error.code === "timeout") {
      timeouts += 1;
    }
  }

  const accepted = input.stageResults.reduce((sum, stage) => sum + Number(stage.submission.accepted ?? 0), 0);
  const expectedAccepts = input.stageResults.reduce((sum, stage) => sum + stage.task_ids.length, 0) || 1;
  let invalidActions = input.plan.normalization_warnings.length;
  if (input.stageResults.some((stage) => stage.error)) {
    invalidActions += 1;
  }

  return {
    plan_validity: 1,
    submit_accept_rate: accepted / expectedAccepts,
    scenario_success: input.verdict.passed ? 1 : 0,
    recovery_success: (input.verdict.passed && input.recoveryExpected) || !input.recoveryExpected ? 1 : 0,
    wall_time_ms: input.wallTimeMs,
    stage_count: input.plan.stages.length,
    task_count: allTasks.length,
    retry_count: retries,
    timeout_count: timeouts,
    runtime_failure_count: runtimeFailures,
    invalid_action_count: invalidActions,
    determinism_drift: 0,
    artifact_hash_match: input.artifactHashMatch ? 1 : 0,
  };
}
